benchmark/benchmark_revisions.py

```diff
@@ -0,0 +1,119 @@
+
+
+def col(text, color):
+    """Wraps the given text in color ANSI sequences.
+
+    Args:
+        text  (str): text to print
+        color (str): ANSI color code
+
+    Returns:
+        str: text wrapped with ANSI codes
+    """
+    return Colors[color] + text + RESET
+
+
+class cd:
+    """Context manager for changing the current working directory"""
+
+    def __init__(self, newPath):
+        self.newPath = os.path.expanduser(newPath)
+        self.savedPath = None
+
+    def __enter__(self):
+        self.savedPath = os.getcwd()
+        os.chdir(self.newPath)
+
+    def __exit__(self, etype, value, traceback):
+        os.chdir(self.savedPath)
+
+
+# benchmarking tool version
+__version__ = "0.1.0"
+
+
+def cli():
+    """Parse the command line arguments, perform the requested action.
+    """
+    parser = argparse.ArgumentParser(description="PennyLane benchmarking tool for revisions")
+    parser.add_argument(
+        "-r",
+        "--revisions",
+        type=lambda x: x.split(","),
+        help="comma-separated list of revisions to run the benchmark on",
+    )
+
+    args, unknown_args = parser.parse_known_args()
+
+    revisions_directory = os.path.join(os.path.dirname(os.path.realpath(__file__)), "revisions")
+    if not os.path.exists(revisions_directory):
+        os.mkdir(revisions_directory)
+
+    for revision in args.revisions:
+        print(">>> Running benchmark for revision {}".format(col(revision, "red")))
+        pl_directory = os.path.join(revisions_directory, revision)
+
+        # We first make sure we get the latest version of the desired revision
+        if os.path.exists(pl_directory):
+            print(">>> Revision found locally, updating...")
+            with cd(pl_directory):
+                # Check if we're on a detached HEAD (i.e. for version revisions)
+                res = subprocess.run(
+                    "git rev-parse --abbrev-ref --symbolic-full-name HEAD",
+                    capture_output=True,
+                    check=True,
+                )
+
+                if "HEAD" not in str(res.stdout):
+                    subprocess.run("git checkout {} -q".format(revision))
+                    subprocess.run("git pull -q")
+        else:
+            print(">>> Revision not found locally, cloning...")
+            os.mkdir(pl_directory)
+            with cd(pl_directory):
+                subprocess.run("git clone https://www.github.com/xanaduai/pennylane . -q")
+                subprocess.run("git checkout {} -q".format(revision))
+
+        benchmark_file_path = os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), "benchmark.py"
+        )
+        benchmark_env = os.environ.copy()
+        benchmark_env["PYTHONPATH"] = pl_directory + ";" + benchmark_env["PATH"]
+        subprocess.run(
+            ["python3", benchmark_file_path] + unknown_args + ["--noinfo"],
+            env=benchmark_env,
+            check=True,
+        )
+
+
+if __name__ == "__main__":
```

Comment:
Nice! Another approach would even be to expose this using `entry points`.

What this would allow, is to be able to use the `cli` as if it was installed as a binary executable. That is a `pl-benchmark` could be added for example with a `revisions` `Parser`. This way once PennyLane has been installed, the `pl-benchmark revisions` command would be available from the terminal.

An example for this was recently added to Strawberry Fields:
* https://github.com/XanaduAI/strawberryfields/blob/a144b4c38034a94853441818d8072a9a9cf41d06/setup.py#L50
* https://github.com/XanaduAI/strawberryfields/blob/master/strawberryfields/cli/__init__.py

Reply:
Something to think about in the future, but it would require making the benchmarking folder a package in PennyLane, so a lot more design work :)

Reply:
Totally for it if it works, but I guess this is overengineering. It will be quite some work to enable this feature and now really give an advantage to the users (i.e. us)